src/libs/API.js

```diff
@@ -191,6 +185,11 @@ Network.registerErrorHandler((queuedRequest, error) => {
     // Set an error state and signify we are done loading
     setSessionLoadingAndError(false, 'Cannot connect to server');
 
+    if (isRequestRetriable(queuedRequest)) {
+        retryRequest(queuedRequest);
+        return;
+    }
```

Comment:
Correct me if wrong, but this here seems to be the main material change? 

If we get a network error (like a fetch error) now we just toss the request into the ether instead of retrying it?

Reply:
If that's correct. I'm sort of confused why we are doing this here instead of when we call `onError()` here

https://github.com/Expensify/App/blob/021ad5d5b057fcecd98ee70b6bf52938cca5fe32/src/libs/Network.js#L226-L228

Could we just add the request to the array of requests to retry instead?

https://github.com/Expensify/App/blob/021ad5d5b057fcecd98ee70b6bf52938cca5fe32/src/libs/Network.js#L243-L245

Seems easier to do that and then we won't need `onError()` to call `retryRequest()` which calls `Network.post()` etc.

Reply:
>The opposite actually. When we get a network/fetch error we'll get right here where we retry if we're supposed to.

Yes, sorry I am describing the **current behavior** (without your changes)

I see your point about losing the logs. But we could also:

- Log that we are retrying because of a network error from within Network
- Add the request back into the queue right away 
- Not call the `onError()` because we don't need to reject the promise or call `setLoadingAndError()`

Reply:
Sounds like a plan

Reply:
@marcaaron This would actually not work

We're in async code here - pushing to `requestsToProcessOnNextRun` would do nothing

```js
.catch(error => {
  if (canRetryRequest(queuedRequest)) {
    requestsToProcessOnNextRun.push(queuedRequest); 
  }
});
```

Even if we make it work (by pushing directly to `networkRequestQueue`) we're introducing a delay of up to one second, until the network queue retries the request

The best thing seems to be to call `.post()` so that the request can be retried immediately

---

Maybe we can extract a `canRetryRequest` with just the logic so that it can be reused in Network and in the error handler in API 
And initiate the retry from the error handler

Instead of having code inline like 

```js
.catch(error => {
  if (canRetryRequest(queuedRequest)) {
    requestsToProcessOnNextRun.push(queuedRequest); 
  } else {
  // ...
});
```

I prefer to have it extracted to a function and the original code already does that

```js
.catch(error => onError(queuedRequest, error));
```

Reply:
Sorry, let's walk through a real example of a request that has fetch failures just to get on the same page? I might be missing the concern because the story in my head is that when fetch fails the network is likely inaccessible.

1. We go offline while the network queue is emptying
1. We get a fetch error
1. We put the request back into the queue 
1. We are offline now so the request queue won't be processed
4. We go online
5. The queue is processed
6. Request succeeds

What is the scenario where we have a fetch error and we would want to trigger an immediate retry?

Reply:
This POV makes sense to me, but I also wonder if we are pre-optimizing for an edge case where a fetch failure happens while we are online. If the request is failing for a reason do it just keep getting tried over and over again? Maybe the once per second will act as a throttle in that case and prevent an infinite network loop (which is a tradeoff when considering the artificial delay mentioned).

This also begs the question of whether we should even retry such things. They failed once while we were online - why should we expect them to succeed the next time? Maybe we should stop trying at some point or if a specific kind of error is caught?

Anyway, I'm finding it difficult to form an opinion without concrete examples of what these errors are. The important thing seems to be that the request is not lost entirely. I am not quite convinced it should be retried indefinitely and somewhat worried we may start to see infinite retry loops without some kind of system in place to prevent them.

Reply:
Another reason to keep things in API is to avoid circular deps we don't have access to logging in Network.js
We can only use console.debug 

In API we can capture a retry and log about it

```js
    if (lodashGet(queuedRequest, 'data.shouldRetry')) {
        Log.hmmm('[API] Request is retriable, retrying...', queuedRequest);
        Network.post(queuedRequest.command, queuedRequest.data, queuedRequest.type, queuedRequest.shouldUseSecure)
```

And the error handler by default captures the fact that we've **handled** an error, while simply putting it back for retry does not

https://github.com/Expensify/App/blob/9c19841e5a1a738311142a3da49ffe74d9dfa20b/src/libs/API.js#L185-L186

Since we're already retrying requests on a few places inside API, wouldn't it be best to extract logic that can cover those as well ([handleExpiredAuthToken](https://github.com/Expensify/App/blob/9c19841e5a1a738311142a3da49ffe74d9dfa20b/src/libs/API.js#L100-L124)) (And all of those are hooked back to the [original requests' promise](https://github.com/Expensify/App/blob/9c19841e5a1a738311142a3da49ffe74d9dfa20b/src/libs/API.js#L168-L171))

Reply:
>Another reason to keep things in API is logging. To avoid circular deps we don't have access to logging in Network.js

That sounds reasonable to me in the short term because it is more work to solve - though this is one of the things we should eventually refactor. We should be able to log in the lib without worrying about a "circular dependency".

>And the error handler by default captures the fact that we've handled an error, while simply putting it back for retry does not

Yes, we would add new logs to say something else in this case like "there was an error and it is being retried"

>Since we're already retrying requests on a few places inside API, wouldn't it be best to extract logic that can cover those as well (handleExpiredAuthToken) (And all of those are hooked back to the original requests' promise)

Didn't really get this, but maybe do not understand. We are not losing the original promise context if we put it back into the request queue with the `resolve` and `reject` attached. If the request gets a response it will be handled.

Reply:
> I am not quite convinced it should be retried indefinitely and somewhat worried we may start to see infinite retry loops without some kind of system in place to prevent them.

Yep, it's possible that a cors error can happen, or an error resulting from calling an url that no longer exists, but these are developer errors and would result in a deploy blocker anyway
- not sure if there's even a chance to call an `url` that doesn't exist - we're calling Expensify's api if an endpoint doesn't exist the server would respond with an error

The requests we retry here would eventually succeed 
But if you're interested in limiting retries to a finite number I have something here: https://github.com/Expensify/App/pull/6556

Reply:
> > Since we're already retrying requests on a few places inside API, wouldn't it be best to extract logic that can cover those as well (handleExpiredAuthToken) (And all of those are hooked back to the original requests' promise)
> 
> Didn't really get this, but maybe do not understand. We are not losing the original promise context if we put it back into the request queue with the `resolve` and `reject` attached. If the request gets a response it will be handled.

I'm saying that we already have a `retry` logic inside API (calling Network.post)
And we're discussing whether we should add another `retry` logic in Network or API (to spare us from calling Network.post)
Let's keep using `Network.post` to retry the request like we do in the Response Handler (expired token)
It would still aid refactoring, when we identify exactly how it should happen

Reply:
>a cors error can happen, or an error resulting from calling an url that no longer exists
>these are developer errors and would result in a deploy blocker anyway

Are these the only possibilities we should expect to see? And if so and these should be deploy blocking events why would the proposed solution be to retry them? Should we maybe at least queue a log to alert in such cases in addition to retrying over and over?

>The requests we retry here would eventually succeed

Sorry, I'm not quite seeing where the eventual success happens.

>Let's keep using Network.post to retry the request like we do in the Response Handler (expired token)

As I mentioned already, I don't see a compelling reason to do it this way so it's not worth having this many back and forth exchanges to resolve.

Regardless, it seems more important to figure out how to handle the infinite retries concern. 

I'm not sure if there's a way to wrap that concern up quickly. But without a proposal to do so I'm not sure it's a great idea to retry these requests.

Reply:
The reason to retry the request, besides everything else said
When we fail to reach the API for any reason - retry the request (if it's retriable) 

Keep the logic simple we can extend it if it becomes a problem

Reply:
>tell me what you need done

Sure, sorry for being indirect - that new plan sounds good for now

>- have the retry happen in Network.js
>- don't log anything
>- push the request directly to the queue

It would be good to log when the request happens so maybe next we can solve the circular dependency issues in a new PR.